pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrTokenManager.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.antlr;
+
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.Lexer;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+
+import net.sourceforge.pmd.cpd.token.GenericAntlrToken;
+
+import com.beust.jcommander.internal.Nullable;
+
+import net.sourceforge.pmd.lang.TokenManager;
+
+/**
+ * Generic token manager implementation for all Antlr lexers.
+ */
+public class AntlrTokenManager implements TokenManager {
+    private final Lexer lexer;
+    private String fileName;
+    private final String commentToken;
+    private GenericAntlrToken previousComment;
+
+    /**
+     * Constructor
+     *
+     * @param lexer The lexer
+     * @param fileName The file name
+     * @param commentToken The list of all comment tokens on the grammar
+     */
+    public AntlrTokenManager(final Lexer lexer, final String fileName, @Nullable final String commentToken) {
+        this.lexer = lexer;
+        this.fileName = fileName;
+        this.commentToken = commentToken;
+        resetListeners();
+    }
+
+    @Override
+    public Object getNextToken() {
+        final Token token = lexer.nextToken();
+        if (isCommentToken(token.getText())) {
+            previousComment = new GenericAntlrToken(token, previousComment);
```

Comment:
this variable is never being reset. All comments end up chained even if not being one immediately after the other